cont.c

```diff
@@ -1427,6 +1369,127 @@ show_vm_pcs(const rb_control_frame_t *cfp,
 }
 #endif
 
+NOINLINE(static void fiber_setcontext(rb_fiber_t *new_fiber, rb_fiber_t *old_fiber));
+
+static void
+fiber_setcontext(rb_fiber_t *new_fiber, rb_fiber_t *old_fiber)
+{
+    rb_thread_t *th = GET_THREAD();
+
+    /* save old_fiber's machine stack - to ensure efficient garbage collection */
+    if (!FIBER_TERMINATED_P(old_fiber)) {
+        STACK_GROW_DIR_DETECTION;
+        SET_MACHINE_STACK_END(&th->ec->machine.stack_end);
+        if (STACK_DIR_UPPER(0, 1)) {
+            old_fiber->cont.machine.stack_size = th->ec->machine.stack_start - th->ec->machine.stack_end;
+            old_fiber->cont.machine.stack = th->ec->machine.stack_end;
+        }
+        else {
+            old_fiber->cont.machine.stack_size = th->ec->machine.stack_end - th->ec->machine.stack_start;
+            old_fiber->cont.machine.stack = th->ec->machine.stack_start;
+        }
+    }
+
+    /* exchange machine_stack_start between old_fiber and new_fiber */
+    old_fiber->cont.saved_ec.machine.stack_start = th->ec->machine.stack_start;
+
+    /* old_fiber->machine.stack_end should be NULL */
+    old_fiber->cont.saved_ec.machine.stack_end = NULL;
+
+    // if (DEBUG) fprintf(stderr, "fiber_setcontext: %p[%p] -> %p[%p]\n", (void*)old_fiber, old_fiber->stack.base, (void*)new_fiber, new_fiber->stack.base);
+
+#if defined(COROUTINE_SANITIZE_ADDRESS)
+    __sanitizer_start_switch_fiber(FIBER_TERMINATED_P(old_fiber) ? NULL : &old_fiber->context.fake_stack, new_fiber->context.stack_base, new_fiber->context.stack_size);
+#endif
+
+    /* swap machine context */
+    struct coroutine_context * from = coroutine_transfer(&old_fiber->context, &new_fiber->context);
+
+#if defined(COROUTINE_SANITIZE_ADDRESS)
+    __sanitizer_finish_switch_fiber(old_fiber->context.fake_stack, NULL, NULL);
+#endif
+
+    if (from == NULL) {
+        rb_syserr_fail(errno, "coroutine_transfer");
+    }
+
+    /* restore thread context */
+    fiber_restore_thread(th, old_fiber);
+
+    // It's possible to get here, and new_fiber is already freed.
+    // if (DEBUG) fprintf(stderr, "fiber_setcontext: %p[%p] <- %p[%p]\n", (void*)old_fiber, old_fiber->stack.base, (void*)new_fiber, new_fiber->stack.base);
+}
+
+static VALUE
+make_passing_arg(int argc, const VALUE *argv)
+{
+    switch (argc) {
+      case -1:
+        return argv[0];
+      case 0:
+        return Qnil;
+      case 1:
+        return argv[0];
+      default:
+        return rb_ary_new4(argc, argv);
+    }
+}
+
+#ifdef RUBY_CALLCC_ENABLED
```

Comment:
```suggestion
#if defined(RUBY_ASAN_ENABLED) || defined(RUBY_MSAN_ENABLED)
# define RUBY_CALLCC_ENABLED 1
#else
# undef RUBY_CALLCC_ENABLED
#endif
```

The other files should not define/use it.

Reply:
So I can do this, but there are parts of `vm_core.h`, `hash.c`, and `vm.c` which are there for callcc support. And I thought if I was going to have a switch which disabled callcc, it really should remove all of the bits of it throughout the codebase.

For example, `hash.c` calls `ruby_register_rollback_func_for_ensure`, and if i don't ifdef it out there, then I need to keep that function in here. And so some parts of callcc support would be compiled out, but some would be kept.

I think the best alternative to this kind of invasive patch would be to simply do

```
#ifdef RUBY_ASAN_ENABLED
    rb_define_global_function("callcc", rb_callcc, 0);
#else
    rb_define_global_function("callcc", rb_f_notimplement, 0);
#end
```

and then go and mark `rb_callcc` with `RBIMPL_ATTR_MAYBE_UNUSED`. But I wanted to check first if we should avoid skipping the warning like that, hence this PR.

Reply:
ASAN is for debugging purpose; it should not change behaviors as possible.

```C
#ifdef RUBY_ASAN_ENABLED
# define rb_callcc ((void)rb_callcc, rb_f_notimplement)
#endif
```